File: CSES/4. Graph Algorithms/CSES-1667_Message Route.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, m, u, v;
    cin >> n >> m;
    vector<vector<int>> g(n);
    for (int i = 0; i < m; i++) {
        cin >> u >> v;
        u--, v--;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    vector<int> pre(n, -1);
    vector<bool> vis(n, false);
    queue<int> q;
    q.push(0);
    vis[0] = true;
    while (!q.empty()) {
        int u = q.front(); q.pop();
        for (int v : g[u]) {
            if (vis[v]) continue;
            vis[v] = true;
            pre[v] = u;
            q.push(v);
        }
    }
    if (pre[n - 1] == -1) {
        cout << "IMPOSSIBLE" << endl;
    } else {
        vector<int> ans;
        for (int u = n - 1; u != -1; u = pre[u]) ans.push_back(u);
        reverse(ans.begin(), ans.end());
        cout << ans.size() << endl;
        for (int u : ans) cout << u + 1 << " ";
        cout << endl;
    }
    return 0;
}

File: CSES/4. Graph Algorithms/CSES-1667_Message Route.py
from collections import deque

n, m = map(int, input().split())

g = [[] for _ in range(n)]
for _ in range(m):
    u, v = map(lambda x: int(x) - 1, input().split())
    g[u].append(v)
    g[v].append(u)

vis = [False] * n
pre = [-1] * n
vis[0] = True
q = deque([0])
while q:
    u = q.popleft()
    for v in g[u]:
        if vis[v]:
            continue
        vis[v] = True
        pre[v] = u
        q.append(v)

if pre[n - 1] == -1:
    print("IMPOSSIBLE")
else:
    ans = []
    x = n - 1
    while x != -1:
        ans.append(x + 1)
        x = pre[x]
    ans.reverse()
    print(len(ans))
    print(*ans)